messaging-activemq/injection/src/main/java/org/wildfly/extension/messaging/activemq/deployment/injection/AbstractJMSContext.java

```diff
@@ -23,12 +23,21 @@
  */
 public abstract class AbstractJMSContext implements Serializable {
 
-    private final Map<String, JMSContext> contexts = new ConcurrentHashMap<>();
+    private final transient ReentrantLock lock = new ReentrantLock();
+    private final transient Map<String, JMSContext> contexts = new HashMap<>();
 
-    JMSContext getContext(String injectionPointId, JMSInfo info, ConnectionFactory connectionFactory) {
-        return contexts.computeIfAbsent(injectionPointId, key -> {
-            return createContext(info, connectionFactory);
-        });
+    synchronized JMSContext getContext(String injectionPointId, JMSInfo info, ConnectionFactory connectionFactory) {
+        lock.lock();
+        try {
+            JMSContext context = contexts.get(injectionPointId);
+            if (context == null) {
+                context = createContext(info, connectionFactory);
+                contexts.put(injectionPointId, context);
+            }
+            return context;
+        } finally {
+            lock.unlock();
+        }
```

Comment:
If this method is called a lot you can stick with ConcurrentHashMap, do a get call, and fall back to this lock stuff if it returns null.

Reply:
@ehsavoie You switched back to ConcurrentHashMap but you're still doing the initial read with the lock held.

Basically I'm suggesting double checked locking.